spark/src/main/scala/org/apache/comet/CometSparkSessionExtensions.scala

```diff
@@ -788,68 +788,93 @@ class CometSparkSessionExtensions
           }
 
         // Native shuffle for Comet operators
-        case s: ShuffleExchangeExec
-            if isCometShuffleEnabled(conf) &&
-              isCometNativeShuffleMode(conf) &&
-              QueryPlanSerde.supportPartitioning(s.child.output, s.outputPartitioning)._1 =>
-          logInfo("Comet extension enabled for Native Shuffle")
+        case s: ShuffleExchangeExec =>
+          val nativePrecondition = isCometShuffleEnabled(conf) &&
+            isCometNativeShuffleMode(conf) &&
+            QueryPlanSerde.supportPartitioning(s.child.output, s.outputPartitioning)._1
 
-          val newOp = transform1(s)
-          newOp match {
-            case Some(nativeOp) =>
-              // Switch to use Decimal128 regardless of precision, since Arrow native execution
-              // doesn't support Decimal32 and Decimal64 yet.
-              conf.setConfString(CometConf.COMET_USE_DECIMAL_128.key, "true")
-              val cometOp = CometShuffleExchangeExec(s, shuffleType = CometNativeShuffle)
-              CometSinkPlaceHolder(nativeOp, s, cometOp)
-            case None =>
-              s
-          }
+          val nativeShuffle: Option[SparkPlan] =
+            if (nativePrecondition) {
+              val newOp = transform1(s)
+              newOp match {
+                case Some(nativeOp) =>
+                  // Switch to use Decimal128 regardless of precision, since Arrow native execution
+                  // doesn't support Decimal32 and Decimal64 yet.
+                  conf.setConfString(CometConf.COMET_USE_DECIMAL_128.key, "true")
+                  val cometOp = CometShuffleExchangeExec(s, shuffleType = CometNativeShuffle)
+                  Some(CometSinkPlaceHolder(nativeOp, s, cometOp))
+                case None =>
+                  None
+              }
+            } else {
+              None
+            }
 
-        // Columnar shuffle for regular Spark operators (not Comet) and Comet operators
-        // (if configured).
-        // If the child of ShuffleExchangeExec is also a ShuffleExchangeExec, we should not
-        // convert it to CometColumnarShuffle,
-        case s: ShuffleExchangeExec
-            if isCometShuffleEnabled(conf) && isCometJVMShuffleMode(conf) &&
+          // this is a temporary workaround because some Spark SQL tests fail
+          // when we enable COMET_SHUFFLE_FALLBACK_TO_COLUMNAR due to valid bugs
+          // that we had not previously seen
+          val tryColumnarNext =
+            !nativePrecondition || (nativePrecondition && nativeShuffle.isEmpty &&
+              COMET_SHUFFLE_FALLBACK_TO_COLUMNAR.get(conf))
```

Comment:
nit: this can be shorter
```
val tryColumnarNext =
    !nativePrecondition || (nativeShuffle.isEmpty && COMET_SHUFFLE_FALLBACK_TO_COLUMNAR.get(conf))
```